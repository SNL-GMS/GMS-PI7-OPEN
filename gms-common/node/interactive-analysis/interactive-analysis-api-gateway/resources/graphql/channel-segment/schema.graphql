enum TimeSeriesType {
    WAVEFORM
    FK_SPECTRA
    DETECTION_FEATURE_MAP
}

# Represents a channel time segment comprised of one or more waveforms for a (derived or raw).
type ChannelSegment {

    # Unique identifier for the channel segment
    id: String!

    # Name of ChannelSegment
    name: String!

    # Type of timeseries
    timeseriesType: TimeSeriesType!

    # The type enumeration of the channel segment
    type: ChannelSegmentType

    # Unique identifier for the channel id
    channelId: String!

    # The start date/time (seconds since epoch) of the channel segment
    startTime: Float!

    # The end date/time ((seconds since epoch) of the channel segment
    endTime: Float!

    # The channel associated with the segment
    channel: ProcessingChannel

    # The collection of Waveforms associated with the channel segment
    timeseries: [Timeseries]

    # Provenance creation information for the channel segment
    creationInfo: CreationInfo!
}

# Input version of the time window parameters provided as input to the FK mutation API
input WindowParametersInput {

    # The type of window (e.g. Hanning)
    windowType: String!

    # The lead period in seconds (prior to the detection onset time)
    leadSeconds: Float!

    # The length of the window in seconds (including the lead and lag around the detection onset time)
    lengthSeconds: Float!

    #  The number of FK Spectra per second to fetch
    stepSize: Float!
}

input ContributingChannelsConfigurationInput {
    id: String!
    enabled: Boolean!
    name: String
}

# Holds the configuration used to calculate an Fk
  
input FkConfigurationInput {
    maximumSlowness: Float!
    mediumVelocity: Float!
    numberOfPoints: Float!
    normalizeWaveforms: Boolean!
    useChannelVerticalOffset: Boolean!
    contributingChannelsConfiguration: [ContributingChannelsConfigurationInput]!
    leadFkSpectrumSeconds: Float!
}
# Input parameters used to create a new FK data set
input FkInput {

    # The ID of the station the FK will be created for
    stationId: String!

    # Signal detection that Fk is created from
    signalDetectionId: String!

    # Signal detection hypothesis
    signalDetectionHypothesisId: String!

    # Phase of the signal detection
    phase: String!

    # The frequency band to compute the Fk on
    frequencyBand: FrequencyBandInput!

    # The window parameters to compute the Fk with
    windowParams: WindowParametersInput!

    # Configuration settings for the continous fk
    configuration: FkConfigurationInput
}

# Input to modify leadSeconds and return SD with leadFkSpectrum and PowerSpectrum populated
input FkLeadInput {
     # Signal detection that Fk is created from
    signalDetectionId: String!

    # Window Lead Seconds and Lead Fk Spectrum Select Seconds (movie)
    windowLeadSeconds: Float! # If lead is changed thru the widget
    leadFkSpectrumSeconds: Float!
}

# Input parameters that indicate which Fk (ChannelSegment TimeSeries)
# to set the reviewed flag
input MarkFksReviewedInput {
    # Signal detection that Fk is created from
    signalDetectionIds: [String]!

    # Boolean flag to set if reviewed by the Analyst
    reviewed: Boolean!
}

extend type Query {
    # Compute the Fk Thumbnails for UI Display. Should be called from UI after a
    # new Fk is called
    computeFkFrequencyThumbnails(fkInput: FkInput!): FkFrequencyThumbnailBySDId!
}

# Currently, no queries are provided. Fk data are accessed as members of
# AzslownessFeatureMeasurements on SignalDetectionHypotheses

# The following signal FK mutations are supported.
extend type Mutation {
    setFkWindowLead(leadInput: FkLeadInput!): AssociationChange

    # Compute the new FKs and set for each signal detection id
    computeFks(fkInput: [FkInput]!): AssociationChange

    # update reviewed flag 
    markFksReviewed(markFksReviewedInput: MarkFksReviewedInput!): AssociationChange

    # Calculate new Beam (replaces ArrivalTime FM Beam)
    computeBeam(fkInput: FkInput!): AssociationChange
}
