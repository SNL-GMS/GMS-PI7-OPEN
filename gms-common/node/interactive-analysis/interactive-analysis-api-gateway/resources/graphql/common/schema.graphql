# Common GraphQL schema definitions shared across the components of the
# API Gateway (e.g. waveform, workflow)

# Units used in DoubleValue part of feature prediction
enum Units {
    DEGREES
    SECONDS
    SECONDS_PER_DEGREE
    UNITLESS
}

# Represents Double fields for a double value representation.
type DoubleValue {
    value: Float # Optional Guidance is value can be a NaN i.e. FP predicted value
    standardDeviation: Float!
    units: Units!
}

# Represents Double fields for a double value representation input from UI.
input DoubleValueInput {
    value: Float!
    standardDeviation: Float!
    units: Units!
}

# Phase types
enum PhaseType {
    P
    S
    P3KPbc
    P4KPdf_B
    P7KPbc
    P7KPdf_D
    PKiKP
    PKKSab
    PKP2bc
    PKP3df_B
    PKSab
    PP_1
    pPKPbc
    PS
    Rg
    SKiKP
    SKKSac
    SKPdf
    SKSdf
    sPdiff
    SS
    sSKSdf
    Lg
    P3KPbc_B
    P5KPbc
    P7KPbc_B
    Pb
    PKKP
    PKKSbc
    PKP2df
    PKPab
    PKSbc
    PP_B
    pPKPdf
    PS_1
    SKKP
    SKKSac_B
    SKS
    SKSSKS
    sPKiKP
    SS_1
    SSS
    nNL
    P3KPdf
    P5KPbc_B
    P7KPbc_C
    PcP
    PKKPab
    PKKSdf
    PKP3
    PKPbc
    PKSdf
    pPdiff
    PPP
    pSdiff
    Sb
    SKKPab
    SKKSdf
    SKS2
    Sn
    sPKP
    SS_B
    SSS_B
    NP
    P3KPdf_B
    P5KPdf
    P7KPdf
    PcS
    PKKPbc
    PKP
    PKP3ab
    PKPdf
    Pn
    pPKiKP
    PPP_B
    pSKS
    ScP
    SKKPbc
    SKP
    SKS2ac
    SnSn
    sPKPab
    sSdiff
    NP_1
    P4KPbc
    P5KPdf_B
    P7KPdf_B
    Pdiff
    PKKPdf
    PKP2
    PKP3bc
    PKPPKP
    PnPn
    pPKP
    PPS
    pSKSac
    ScS
    SKKPdf
    SKPab
    SKS2df
    SP
    sPKPbc
    sSKS
    P4KPdf
    P5KPdf_C
    P7KPdf_C
    Pg
    PKKS
    PKP2ab
    PKP3df
    PKS
    PP
    pPKPab
    PPS_B
    pSKSdf
    Sdiff
    SKKS
    SKPbc
    SKSac
    SP_1
    sPKPdf
    sSKSac
    Sx
    tx
    N
    Px
    PKhKP
    LR
    pP
    UNKNOWN
}

# A date/time Type
scalar Date

# Represents a time range
# type TimeRange {
#     startTime: Float!
#     endTime: Float!
# }

# Input verion of a time range
input TimeRange {
    startTime: Float!
    endTime: Float!
}

# Represents a location specified using latitude (degrees), longitude (degrees),
# and altitude (kilometers).
type Location {

    # Latitude in degrees
    latDegrees: Float!
    
    # Longitude in degrees
    lonDegrees: Float!

    # Elevation in kilometers
    elevationKm: Float!
}

input LocationInput {

    # Latitude in degrees
    latDegrees: Float!
    
    # Longitude in degrees
    lonDegrees: Float!

    # Elevation in kilometers
    elevationKm: Float!
}

# Distance value's units degrees or kilometers
enum DistanceUnits {
    degrees
    km
}

# Enumerated list of source types used to compute distances to
enum DistanceSourceType {
    Event
    UserDefined
}

# Represents a distance measurement relative to a specified source location
type DistanceToSource {

    # The distance in kilometers or degrees to the source
    distance: Float!

    # The units distance was calculated
    distanceUnits: DistanceUnits

    # The source location
    sourceLocation: Location!

    # The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType!

    # Optional: the unique ID of the source object
    sourceId: String!

    # Which station distance to the source
    stationId: String!
}

input DistanceToSourceInput {
    # The source location
    sourceLocation: LocationInput

    # The type of the source the distance is measured to (e.g. and event)
    sourceType: DistanceSourceType

    # The Distance Units to use
    distanceUnits: DistanceUnits

    # Optional: the unique ID of the source object
    sourceId: String
}

# Represents a frequency range
type FrequencyBand {
    # Minimum frequency in Herz
    minFrequencyHz: Float!

    # Maximum frequency in Herz
    maxFrequencyHz: Float!
}

# Represents a frequency range input to a mutation
input FrequencyBandInput {
    # Minimum frequency in Herz
    minFrequencyHz: Float!

    # Maximum frequency in Herz
    maxFrequencyHz: Float!
}

# Enumeration representing the different types of stations in the monitoring network
enum StationType {

    Seismic3Component,
    Seismic1Component,
    SeismicArray,
    Hydroacoustic,
    HydroacousticArray,
    Infrasound,
    InfrasoundArray,
    Weather,
    UNKNOWN
}

# Type enumeration for channel segments
enum ChannelSegmentType {
    ACQUIRED
    RAW
    DETECTION_BEAM
    FK_BEAM
    FILTER
    FK_SPECTRA
}

# Provenance reference details for information added to the System
type InformationSource {

    # Unique ID for the source
    id: String!

    # Source organization
    organizationName: String!

    # The time associated with the information source
    referenceTime: Float!
}

# Enumeration of configuration status changes reflected in the configuration
enum ConfigurationInfoStatus {

    # Initial configuration
    Initial

    # The system's installed location for the item associated with the configuration was corrected
    CorrectedSystemLocation

    # The physical location of the ite associated with the configuration was changed
    MovedPhysicalLocation

    # The item associated with the configuration was decommissioned
    Decommissioned
}

# Provenance information associated with a configuration item in the system
type ConfigurationInfo {

    # The unique ID of the user who created the Configuration
    creatorId: String!

    # The time the configuration took effect in the real world (e.g. station came online)
    effectiveTime: Float!

    # The time the configuration was installed (and thus took effect) in the System
    installationTime: Float!

    # The status change reflected in the configuration values
    status: ConfigurationInfoStatus!

    # Comment associated with the configuration information
    comment: String
}

# Enumerated type of the actor (e.g. analyst, system) associated with a CreationInfo object
enum CreatorType {
    Analyst
    System
}

# Provenance information about the results of data processing completed by the System
# and actions completed by users of the system
type CreationInfo {
    # The unique ID of the creation info object
    id: String!

    # The time (seconds since Unix epoch) the associated processing result was created
    creationTime: Float!

    # The unique ID of the user who created the associated processing result
    creatorId: String!

    # The enumerated type of the creator (e.g. analyst, system)
    creatorType: CreatorType!

     # The name of the creation
    creatorName: String!

    # TODO
    # The processing step associated with the processing result
    # processingStep: ProcessingStep!

    # TODO
    # The software component associated with the processing result
    # softwareComponent: SoftwareComponent!
}

input CreationInfoInput {
    # Identifier for the creation info
    creatorName: String!
    
    # The time (seconds since Unix epoch) the associated processing result was created
    creationTime: Float!

    # The software information associated with the processing result
    softwareInfo: SoftwareInfoInput!
}

type SoftwareInfo {
    # Identifier for the creation info
    name: String!

    # The software version identifier
    version: String!
}

input SoftwareInfoInput {
    # Identifier for the creation info
    name: String!

    # The software version identifier
    version: String!
}

input ProcessingContextInput {
    analystActionReference: AnalystActionReferenceInput!
    processingStepReference: ProcessingStepReferenceInput!
    storageVisibility: String!
}

input AnalystActionReferenceInput {
    processingStageIntervalId: String!
    processingActivityIntervalId: String!
    analystId: String!
}

input ProcessingStepReferenceInput {
    processingStageIntervalId: String!
    processingSequenceIntervalId: String!
    processingStepId: String!
}

# Represents an alias for a station, channel, site, etc.
type Alias {

    # The alias name
    name: String!

    # Provenance information about the alias
    configurationInfo: ConfigurationInfo
}

# Represents calibration information associated with a waveform
type ProcessingCalibration {

    # Calibration factor
    factor: Float!

    # Calibration factor error
    factorError: Float!

    # Calibration period
    period: Float!

    # Calibration time shift
    timeShift: Float!
}

# Input version of calibration information associated with a waveform
input ProcessingCalibrationInput {

    # Calibration factor
    factor: Float!

    # Calibration factor error
    factorError: Float!

    # Calibration period
    period: Float!

    # Calibration time shift
    timeShift: Float!
}

# Represents a generic timeseries interface containing the fields common to the
# concrete timeseries data types defined in the system (e.g. Waveform, FkPowerSpectrum, DetectionFeatureMap)
interface Timeseries {
    # The start date/time (seconds since epoch) of the timeseries
    startTime: Float!

    # The sample rate (Hz) of the samples in the timeseries
    sampleRate: Float!

    # The sample count in the timeseries
    sampleCount: Float!
}

input TimeseriesInput {
    # The start date/time (seconds since epoch) of the timeseries
    startTime: String!
    
    # The sample rate (Hz) of the samples in the timeseries
    sampleRate: Float!
    
    # The sample count in the timeseries
    sampleCount: Float!
}

# Represents a time-bounded segment of waveform samples output from a channel (derived or raw).
# Waveforms implement the Timeseries interface
type Waveform implements Timeseries {
    # Unique identifier for the waveform
    # id: String!

    # The start date/time (seconds since epoch) of the waveform
    startTime: Float!

    # The sample rate (Hz) of the samples in the waveform
    sampleRate: Float!

    # The sample count in the waveform
    sampleCount: Float!

    # The time-ordered collection of waveform sample values
    values: [Float]
}

# Input version of a time-bounded segment of waveform samples output from a channel (derived or raw).
# Waveforms implement the Timeseries interface
input WaveformInput {
    
    # Unique identifier for the waveform
    id: String!

    # The start date/time (seconds since epoch) of the waveform
    startTime: Float!

    # The end date/time (seconds since epoch) of the waveform
    endTime: Float!    # The sample rate (Hz) of the samples in the waveform
    sampleRate: Float!

    # The sample count in the waveform
    sampleCount: Float!

    # The time-ordered collection of waveform sample values
    values: [Float]

    # Calibration information for the waveform
    calibration: ProcessingCalibrationInput
}

# The following queries are supported
type Query {
    # Dummy
    queryCount: Int
}

# The following mutations are supported
type Mutation {
    # Dummy
    mutationCount: Int
}

# The following subscriptions are supported
type Subscription {
    # Dummy
    subscriptionCount: Int
}

# Association Change object to return
type AssociationChange {
    events: [Event]
    sds: [SignalDetection]
}

 # Used when we need to return the original event mutated where performed on
 # in addition to the other data changed
type EventAndAssociationChange {
    event: Event!
    associationChange: AssociationChange!
}

#
enum LogLevel {
    INFO
    DEBUG
    WARNING
    ERROR
    DATA
}

type ClientLog {
    logLevel: LogLevel
    message: String!
    time: String
}

input ClientLogInput {
    logLevel: LogLevel!
    message: String!
    time: String
}

extend type Mutation {
    clientLog(clientLogInput: ClientLogInput): ClientLog
}