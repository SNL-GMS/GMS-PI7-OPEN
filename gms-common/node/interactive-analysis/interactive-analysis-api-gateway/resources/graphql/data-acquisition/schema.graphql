
# Represents the transfer gap fields used for gap status.
type FileGap {
    # Station Id
    stationName: String

    # Channel Id
    channelNames: [String]

    # Start time for transfer gap
    startTime: String

    # End time for transfer gap
    endTime: String

    # Duration for transfer gap
    duration: String

    # Location for transfer gap
    location: String

    # Priority for transfer gap
    priority: String
}

# The following queries are supported.
extend type Query {
    # Retrieve transfer file information by time range
    transferredFilesByTimeRange(timeRange: TimeRange!): [FileGap]
}

# Input to Reference Station mutations
input ReferenceStation {
    name: String!
    description: String!
    stationType: StationType!
    comment: String!
    source: InformationSourceInput!
    latitude: Float!
    longitude: Float!
    elevation: Float!
    actualChangeTime: String!
    systemChangeTime: String!
    aliases: [ReferenceAliasInput]!
}

input InformationSourceInput {
    originatingOrganization: String
    informationTime: String
    reference: String
}

input ReferenceAliasInput {
    id: String
    name: String
    status: StatusTypeName
    comment: String
    actualChangeTime: String
    systemChangeTime: String
}

enum StatusTypeName {
    INACTIVE,
    ACTIVE
}

type SaveStationResult{
    id: String!
    result: Boolean!
}

# The following Data Aquisition mutations are supported.
extend type Mutation {
    saveReferenceStation (input: ReferenceStation!): SaveStationResult
}