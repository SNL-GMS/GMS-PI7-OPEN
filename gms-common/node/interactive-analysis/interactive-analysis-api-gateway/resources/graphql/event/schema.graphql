# Enumerated status of an event
enum EventStatus {

    # The event is ready to be opened for refinement
    ReadyForRefinement

    # The event is open for refinement
    OpenForRefinement

    # The event is awaiting review
    AwaitingReview

    # The event is marked as complete
    Complete
}

# Represents an event marking the occurrence of some transient
# source of energy in the ground, oceans, or atmosphere 
type Event {

    # The unique ID of the event
    id: String!

    # The monitoring organization associated with the event
    monitoringOrganization: String

    # The preferred hypothesis of the event for the provided
    # processing stage
    preferredHypothesisForStage (forStage: ForStage!): PreferredEventHypothesis

    # The current event hypothesis (UI sources values)
    currentEventHypothesis: PreferredEventHypothesis!

    # The history of preferred hypotheses for the event, ordered from
    # earliest to latest
    preferredHistory: [PreferredEventHypothesis]!

    # The list of preferred hypotheses for the event associated with 
    # each processing stage
    preferredForStageHistory: [PreferredEventHypothesis]!

    # The collection of hypotheses associated with the event
    hypotheses: [EventHypothesis]!

    # The current state of the event
    status: EventStatus

    # The list of currently-active analysts for the event
    activeAnalysts: [Analyst]

    # Modified flag indicates something in the event has changed and needs saving
    modified: Boolean!

    # Flag that indicates if the event has an associated SD Hyp that is also associated to another event
    hasConflict: Boolean!

    # List of Signal Detection hypotheses with conflict
    conflictingSdHypIds: [String]!
}

# Input for queries by processing stage
input ForStage {
    stageId: String!
}

# The preferred hypothesis for the event at a given processing stage
type PreferredEventHypothesis {

    # The processing stage that the event hypothesis is preferred for
    processingStage: ProcessingStage

    # The preferred event hypothesis for the stage
    eventHypothesis: EventHypothesis!
}

# Represents a proposed explanation for an event, such that the set of
# event hypotheses grouped by an Event represents the history of that event.
type EventHypothesis {

    # The unique ID of the event hypothesis
    id: String! # ! must come back

    # Flag indicating whether or not the event hypothesis has been rejected
    rejected: Boolean!

    # The event this hypothesis pertains to
    event: Event!

    # The max arrival time is set so we don't have to send all the feature
    # measurements for each SignalDetectionEventAssociation below.
    # max arrival time value is set in the resolver for the signalDetectionAssociations
    associationsMaxArrivalTime: Float!
    
    # The signal detection hypothesis associations for the event hypothesis
    signalDetectionAssociations: [SignalDetectionEventAssociation]!

    # The location solutions associated with the event hypothesis
    # multiple solutions are formed using different location constraints such as depth
    # (i.e. unconstrained, fixed to surface, fixed to a depth below the surface), location, or time.
    locationSolutionSets: [LocationSolutionSet]! # this is a required array

    # The preferred location solution for the event hypothesis
    preferredLocationSolution: PreferredLocationSolution
}

type SignalDetectionEventAssociationInput {

    # The unique ID of the association
    id: String!

    # The associated signal detection hypothesis
    signalDetectionHypothesisId: String!

    # The associated event hypothesis
    eventHypothesisId: String!

    # Flag indicating whether or not the association has been rejected
    rejected: Boolean!
}

# Represents the linkage between Event Hypotheses and Signal Detection Hypotheses.
# The rejected attribute is used to ensure that any rejected associations will not
# be re-formed in subsequent processing stages.
type SignalDetectionEventAssociation {

    # The unique ID of the association
    id: String!

    # The associated signal detection hypothesis
    signalDetectionHypothesis: SignalDetectionHypothesis!

    # The associated event hypothesis
    eventHypothesisId: String!

    # Flag indicating whether or not the association has been rejected
    rejected: Boolean!
}

# Location Solution Set
# defines a list of location solutions for an event hypotheis
# including a snapshot of association when solutions were created
type LocationSolutionSet {
    id: String!
    count: Int!
    locationSolutions: [LocationSolution]!
}

# Snapshot of state of associations when location solution was created
type SignalDetectionSnapshot {
    signalDetectionId: String!
    signalDetectionHypothesisId: String!
    stationId: String!
    stationName: String!
    channelName: String!
    phase: String!
    time: EventSignalDetectionAssociationValues!
    slowness: EventSignalDetectionAssociationValues!
    azimuth: EventSignalDetectionAssociationValues!
}

# Generic interface for snapshot values of a signal detection association
type EventSignalDetectionAssociationValues {
    defining: Boolean!
    observed: Float!
    residual: Float
    correction: Float
}

# Represents an estimate of the location of an event, defined as latitude, longitude, depth, and time.
# A location solution is often determined by a location algorithm that minimizes the difference between
# feature measurements (usually arrival time, azimuth, and slowness) and corresponding feature predictions.
type LocationSolution {
    id: String!
    location: EventLocation!
    featurePredictions: [FeaturePrediction]!
    locationRestraint: LocationRestraint!
    locationUncertainty: LocationUncertainty
    locationBehaviors: [LocationBehavior]!
    locationType: String!
    snapshots: [SignalDetectionSnapshot]!
}
# Restraint used in location calcs
type LocationRestraint {
    depthRestraintType: String!
    depthRestraintKm: Float
    latitudeRestraintType: String!
    latitudeRestraintDegrees: Float
    longitudeRestraintType: String!
    longitudeRestraintDegrees: Float
    timeRestraintType: String!
    timeRestraint: String
}

type LocationUncertainty {
    xy: Float!
    xz: Float!
    xt: Float!
    yy: Float!
    yz: Float!
    yt: Float!
    zz: Float!
    zt: Float!
    tt: Float!
    stDevOneObservation: Float!
    ellipses: [Ellipse]
    ellipsoids: [Ellipsoid]
}
type Ellipse {
    scalingFactorType: String!
    kWeight: Float!
    confidenceLevel: Float!
    majorAxisLength: String!
    majorAxisTrend: Float!
    minorAxisLength: String!
    minorAxisTrend: Float!
    depthUncertainty: Float!
    timeUncertainty: String!
}

type Ellipsoid {
    scalingFactorType: String!
    kWeight: Float!
    confidenceLevel: Float!
    majorAxisLength: String!
    majorAxisTrend: Float!
    majorAxisPlunge: Float!
    intermediateAxisLength: Float!
    intermediateAxisTrend: Float!
    intermediateAxisPlunge: Float!
    minorAxisLength: String!
    minorAxisTrend: Float!
    minorAxisPlunge: Float!
    depthUncertainty: Float!
    timeUncertainty: String!
}

# LocationBehavoir type and input
type LocationBehavior {
    residual: Float!
    weight: Float!
    defining: Boolean!
    featurePredictionId: String!
    featureMeasurementId: String!
}

input LocationBehaviorInput {
    residual: Float!
    weight: Float!
    defining: Boolean!
    featurePredictionId: String!
    featureMeasurementId: String!
}

type EventLocation {
  latitudeDegrees: Float!
  longitudeDegrees: Float!
  depthKm: Float!
  time: Float!
}
# Enumerated type of magnitude solution (surface wave, body wave, local, etc.)
#enum MagnitudeType {
#    mb
#    mbMLE
#    mbrel
#    ms
#    msMLE
#    msVMAX
#    ml
#}

# An estimate of the event's magnitude based on detections from multiple stations
#type NetworkMagnitudeSolution {

    # The type of magnitude estimate (e.g. surface wave, body wave, local, etc.)
#    magnitudeType: MagnitudeType!

    # The magnitude value
#    magnitude: Float!
#}

# Represents a preference relationship between an event hypothesis and a location solution.
# Creation information is included in order to capture provenance of the preference.
type PreferredLocationSolution {

    # The preferred location solution
    locationSolution: LocationSolution!

    # Creation information for the preference relationship
    creationInfo: CreationInfo!
}

# Encapsulates input used to create a location solution for an event hypothesis
input EventLocationInput {

    # Latitude in degrees
    latitudeDegrees: Float!
    
    # Longitude in degrees
    longitudeDegrees: Float!

    # Depth in kilometers
    depthKm: Float!

    # Time (seconds since Unix epoch)
    time: Float!
}

# Represents a Feature Prediction as part of the Location Solution. This should represent a 
# predicted location of the event.
type FeaturePrediction {
    id: String
    phase: String
    predictedValue: FeatureMeasurementValue!
    featurePredictionComponents: [FeaturePredictionComponent]! # Java declares as a set not sure in JSON
    extrapolated: Boolean!
    predictionType: FeatureMeasurementTypeName
    sourceLocation: EventLocation!
    receiverLocation: Location!
    channelId: String
    stationId: String
}

# Feature Prediction Component definition
type FeaturePredictionComponent {
    value: DoubleValue!
    extrapolated: Boolean!
    predictionComponentType: FeaturePredictionCorrectionType!
}

# Enumerated types for the Feature Prediction Corrections
enum FeaturePredictionCorrectionType {
    BASELINE_PREDICTION
    ELEVATION_CORRECTION
    ELLIPTICITY_CORRECTION
}

# Encapsulates input used to create a new event hypothesis.
input CreateEventHypothesisInput {

    # The IDs of signal detection hypothesis associated with the event hypothesis
    associatedSignalDetectionIds: [String]!

    # The location solution input for the event hypothesis
    eventLocationInput: EventLocationInput

    # The ID of the actor who is creating the event hypothesis (e.g. analyst userId)
    creatorId: String!

    # The ID of the processing stage the event hypothesis is being created in
    processingStageId: String!
}

# Encapsulates input parameters used to update an existing event hypothesis in line.
input UpdateEventHypothesisInput {

    # Flag indicating whether or not the event hypothesis has been rejected
    rejected: Boolean

    # The IDs of signal detection hypothesis associated with the event hypothesis
    associatedSignalDetectionIds: [String]

    # The location solution input for the event hypothesis
    eventLocationInput: EventLocationInput

    # The ID of the actor who is creating the event hypothesis (e.g. analyst userId)
    creatorId: String!

    # The ID of the processing stage the event hypothesis is being created in
    processingStageId: String!
}

# Encapsulates input parameters used to update an existing event (without the need for
# a new event hypothesis)
input UpdateEventInput {

    # The ID of the actor who is creating the event hypothesis (e.g. analyst userId)
    creatorId: String!

    # The ID of the processing stage the event hypothesis is being created in
    processingStageId: String!

    # The event status to update (optional)
    status: EventStatus

    # The preferred event hypothesis ID (optionsl)
    preferredHypothesisId: String

    # The list of currently-active analysts for the event
    activeAnalystUserNames: [String]
}

# The following event related queries are supported.
extend type Query {
    eventHypothesesInTimeRange(timeRange: TimeRange!): [EventHypothesis]
    eventsInTimeRange(timeRange: TimeRange!): [Event]
    eventById(eventId: String!): Event
}

# The following event related mutations are supported.
extend type Mutation {
    # Creates a new event and associates the SD Hyps passed in
    # SD Hyps are not required, an empty event can be created
    createEvent (signalDetectionHypoIds: [String]): EventAndAssociationChange

    # Update the list of existing events event for cases where new event
    # hypotheses will not be created (e.g. event status)
    updateEvents (eventIds: [String]!, input: UpdateEventInput!): [Event]
    
    # Create a new list of event hypotheses based on the input
    createEventHypotheses (eventIds: [String]!, input: CreateEventHypothesisInput!): [EventHypothesis]

    # Update an existing list of event hypotheses in place using the provided input
    # (i.e. without creating a new event hypotheses)
    updateEventHypothesis (hypothesisIds: [String]!, input: UpdateEventHypothesisInput!): EventHypothesis

    # Call to recompute the feature prediction for event based on input params (TBD)
    updateFeaturePredictions(eventId: String!): Event

    # Mutation to (un)associate Signal Detections to event hypothesis
    changeSignalDetectionAssociations(eventHypothesisId: String!, signalDetectionHypoIds:[String]!, associate: Boolean!): AssociationChange

    # Mutation to Locate Event
    locateEvent(eventHypothesisId: String!, preferredLocationSolutionId: String!, locationBehaviors: [LocationBehaviorInput]!): EventHypothesis
},

# The following event-related subscriptions are supported.
extend type Subscription {
    # Subscribe for events that have been saved to OSD
    # via the save event mutation.
    eventsCreated: [Event]
}
