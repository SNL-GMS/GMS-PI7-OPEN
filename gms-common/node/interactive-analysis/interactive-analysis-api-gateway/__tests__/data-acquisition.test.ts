import { graphql } from 'graphql';
import { schema } from '../src/ts/schema';
import { delayExecution } from '../src/ts/util/delay-execution';

const delayMs = 500;

// ---- Query test cases ----

// Test case - basic content query for data acquisition
// Query for transferred files by time range
it('Transferred Files Content query should match snapshot', async () => {
  // language=GraphQL
  const query = `
  query transferredFilesByTimeRange {
    transferredFilesByTimeRange(timeRange: {startTime: 0, endTime: 9999999999}) {
      stationName
      channelNames
      startTime
      endTime
      duration
      location
      priority
    }
  }
  `;

  const rootValue = {};
  // Execution the GraphQL query with a small delay to allow the API gateway to settle async
  // HTTP requests fetching data from the mock backend
  const result = await delayExecution(() => graphql(schema, query, rootValue), delayMs);
  const { data } = result;

  // Compare response to snapshot
  expect(data).toMatchSnapshot();
});
