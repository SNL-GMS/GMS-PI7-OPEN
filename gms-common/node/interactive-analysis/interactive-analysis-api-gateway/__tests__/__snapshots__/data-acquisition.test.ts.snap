// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transferred Files Content query should match snapshot 1`] = `
Object {
  "transferredFilesByTimeRange": Array [
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "4.01",
      "endTime": "2018-12-19T11:36:15.319925Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T11:36:11.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "5",
      "endTime": "2018-12-19T22:36:11.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:06.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "5",
      "endTime": "2018-12-19T22:36:11.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:06.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "5",
      "endTime": "2018-12-19T22:36:11.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:06.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "5",
      "endTime": "2018-12-19T22:36:11.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:06.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "4",
      "endTime": "2018-12-19T22:36:15.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:11.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "4",
      "endTime": "2018-12-19T22:36:15.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:11.309924Z",
      "stationName": "ASAR",
    },
    Object {
      "channelNames": Array [
        "AS01/SHZ",
      ],
      "duration": "4",
      "endTime": "2018-12-19T22:36:15.309924Z",
      "location": "location",
      "priority": "testPriority",
      "startTime": "2018-12-19T22:36:11.309924Z",
      "stationName": "ASAR",
    },
  ],
}
`;
